cocotb/types/__init__.py

```diff
@@ -0,0 +1,30 @@
+# Copyright cocotb contributors
+# Licensed under the Revised BSD License, see LICENSE for details.
+# SPDX-License-Identifier: BSD-3-Clause
+from .range import Range  # noqa: F401
+from .array import Array  # noqa: F401
+
+
+def concat(a: Array, b: Array) -> Array:
+    """
+    Create a new array that is the concatenation of one array with another.
+
+    Uses the :meth:`__concat__` or :meth:`__rconcat__` special methods to dispatch to a particular implementation,
+    exactly like other binary operations in Python.
+
+    Raises:
+        TypeError: when the arguments do not support concatenation in the given order.
+    """
+    if hasattr(a, '__concat__'):
```

Comment:
There are situations where `Rrconcat` should be tried first: see https://github.com/brettcannon/desugar/blob/290bf659bfc054af3979a9ab29003db6f855e959/desugar/operator.py#L75

Reply:
Ah, I think I understand. If the right hand operation is a subtype of the left hand and the `__rconcat__` is not the same function, it's likely something more specific. The whole `__op__`, `__rop__` business being a poor man's multiple dispatch, where in such a system the runtime would pick the most specific implementation, the second in the example being the most specific if `concat(Array, Unsigned)` were called.

```
concat(Array, Array) -> Array
concat(Array, Unsigned) -> Array  # Unsigned repr is an int here, so we have to do some extra work if we are concatenating with an Array
```

Reply:
Addressed in the most recent commit.